### ISSUE TITLE
How to use library under Shadow DOM?

### ISSUE THREAD

**Issue by chan-dev** [stephanrauh/ngx-extended-pdf-viewer#900]
**Is your feature request related to a problem? Please describe.**
We're displaying some of our components inside Shadow DOM thru `ViewerEncapsulation.ShadowDOM`, and since those components are under shadowdom, it seems like the underlying calls in the library are only using DOM queries.

The component where the template below has a encapsulation of `ViewerEncapsulation.ShadowDOM`
```
<ngx-extended-pdf-viewer [src]="pdfPath" height="100vh" [useBrowserLocale]="true"></ngx-extended-pdf-viewer>
```

We're having issues such as this error:
![pdfviewer](https://user-images.githubusercontent.com/46891115/132075246-034682fc-89d5-4528-934c-281c420c8bea.png)

**Describe the solution you'd like**
Allow the library to work even under Shadow DOM. Provide a `@Input` where you can set a container? In that way, users can set an element container. 

I don't know if something like this is possible.

```
<ngx-extended-pdf-viewer
  [container]="shadowRootElem"
  [src]="'assets/sample.pdf'"
  height="100vh"
  [useBrowserLocale]="true"
></ngx-extended-pdf-viewer>

```

And then inside your class, you can
```
shadowRootElem = document.querySelector('some-element').shadowRoot.querySelector('some-element-under-shadow-dom')
```

Basically, updating some dom queries inside the library.
For instance, the document query inside `needsDummyWidget`

```
needsDummyWidget(id) {
        const widget = document.getElementById(id);
        if (!widget) {
            return true;
        }
        return false;
    }
```

can be written as
```
needsDummyWidget(id) {
       // myContainer could be something passed via @Input container
       // myContainer = document.querySelector('app-root').shadowRoot.querySelector('#toolbarViewer')
        const widget = myContainer.querySelector(`$#{id}`);
        if (!widget) {
            return true;
        }
        return false;
    }
```

**Comment by stephanrauh** [comment]
Makes sense. But I'm afraid it's a major project. I've found 127 usages of `document.getElement*` in the base library pdf.js. So we'd have to develop a strategy how to modify the source code automatedly.

BTW, if my quick research is correct, there's a simpler solution. Replacing `document.getElementById()` by `someLocalDiv.querySelector()` does the trick just as well, assuming that `someLocalDiv` is inside the shadow DOM. Can you confirm this?

**Comment by chan-dev** [comment]
Yup, as long as you've got a reference to the DOM under shadow dom, you can use querySelector/querySelectorAll methods.

**Comment by Morente5** [comment]
Is this being considered?
Thanks a lot.

**Comment by stephanrauh** [comment]
@Morente5 Sorry for the sluggish response. COVID-19 kept me busy. 

As for your question: yes, I still would like to investigate the topic, but it's low priority because few Angular developers are using the shadow DOM. These days I have a hard time just to keep up with the flood of new tickets, so I have to triage. More often than not, the interesting tickets (such as this one) get the shortest straw.

**Comment by dlf06j** [comment]
i need this :(

**Comment by stephanrauh** [comment]
I'm surrending to the flood of issues. Supporting shadow DOM is one of my pet issues, but these days, my library is so popular I barely manage to answer new tickets - let alone solving them. So I reluctantly close this ticket.

To everybody who reads this and / or needs this feature: can I talk you into contributing to my library? I believe this ticket is one of the easy ones. I'll give you the support you need to solve it. Does this sound attractive?

Best regards,
Stephan

**Comment by Morente5** [comment]
> I'm surrending to the flood of issues. Supporting shadow DOM is one of my pet issues, but these days, my library is so popular I barely manage to answer new tickets - let alone solving them. So I reluctantly close this ticket.
> 
> To everybody who reads this and / or needs this feature: can I talk you into contributing to my library? I believe this ticket is one of the easy ones. I'll give you the support you need to solve it. Does this sound attractive?
> 
> Best regards, Stephan

Sure, I would love to contribute.
How do we get in contact and what are the next steps?

**Comment by stephanrauh** [comment]
@Morente5 Awesome! Drop me an email: webmaster1 at beyondjava dot de.

Si prefieres, ¡escríbame en español!

**Comment by ferzamo** [comment]
> @Morente5 Awesome! Drop me an email: webmaster1 at beyondjava dot de.
> 
> Si prefieres, ¡escríbame en español!

Was any solution finally found or developed on this issue?

**Comment by stephanrauh** [comment]
Not that I know of. I'm still drowning in a flood of issues, so I didn't manage to look into the topic myself.

**Comment by en-nicolabaesso** [comment]
> Not that I know of. I'm still drowning in a flood of issues, so I didn't manage to look into the topic myself.

Hi, are you still looking for a PR on this issue (on this library and/or your pdf.js fork)?
I would like to contribute on this, we (=me and my team) are using the library and we could have a use case for this.

**Comment by stephanrauh** [comment]
@en-nicolabaesso Oh yes, that would be great. These days, my thoughts revolve a related topic: supporting multiple viewers on the same page. Both enhancement share a lot of common tasks. If you're working on one of them, you're also working on the other task. Sounds like a win-win situation to me!

One quick word of caution: I'm not entirely sure it's possible to support shadow DOM. Potential problems might be font loading and the web worker. Claude AI also warns about a potential performance penalty.

Be that as it may, it's a major task, so I'd like to split it into multiple pull requests. Smaller PRs are easier to check and merge. Plus, I'm convinced every partial PR is progress.

Here's a sketch of a plan:

- Inject the CSS code into the shadow DOM. You're lucky: I'm already injecting the CSS programmatically. So "all" you have to do is to check if the viewer is running in the shadow DOM, and to inject the CSS into the shadow root. That's handled in the TypeScript classs `DynamicCssComponent`.
- Modify the TypeScript code to deal with the shadow DOM and / or multiple instances of the viewer. There are quite a few `getElementByID()` and `querySelector()` calls in the code. Modify them to run relative to the viewer root instead of using `document.getElementById()`.
- Modify my fork of pdf.js. Use the "bleeding edge" branch as a basis. pdf.js also has quite a few calls of `getElementByID()` and `querySelector()`. They must be modify to search the viewer root instead of scanning the entire document.
  - To do so, pass the root DOM element of the viewer to the class `PdfViewerApplication`. You'll also want to propagate it to the objects created by `PdfViewerApplication` that use `document.getElementById()` or `document.querySelector()`. The good news: that's merely a handful of classes.
  - Replace all calls of document.getElementById()` or `document.querySelector()` by relative calls, with the possible exception of the `PrintService`.
  - Most calls of `document.getElementById()` are in the function `getViewerConfiguration()`. That function is (transitively) called from the TypeScript code, so all you have to do is passing the shadow DOM root and replacing `document.getElementId()` by a query relative to the shadow DOM root.
- pdf.js and ngx-extended-pdf-viewer communicate a lot via an `eventBus`. That's a global object. We have to modify the event bus such that events are delivered to the intended instance of the viewer (in the multiple-viewers scenario) and that they update the DOM inside the shadow root. I'm not sure how to do this, but I suppose that'll be easy once we get there.
- Finally, Claude AI tells me we have to modify the module of the library. I'm not sure about it - especially in the light of the upcoming migration to stand-alone components. My gut tells me that's nonsense, but you never now, so let's add the suggested source code. 

```ts
@NgModule({
  // ... otras configuraciones
  providers: [
    {
      provide: DOCUMENT,
      useFactory: (injector: Injector) => {
        // Detectar si estamos en shadow DOM y ajustar accordingly
        return document;
      },
      deps: [Injector]
    }
  ]
})
export class NgxExtendedPdfViewerModule {
  static forRoot(config?: PdfViewerConfig): ModuleWithProviders<NgxExtendedPdfViewerModule> {
    return {
      ngModule: NgxExtendedPdfViewerModule,
      providers: [
        { provide: PDF_VIEWER_CONFIG, useValue: { ...config, shadowDomSupport: true } }
      ]
    };
  }
}
````

I think that's it in a nutshell. It sounds a lot, and it is - but I think if we do it step by step, in small chunks, we'll manage!
